Prototype

kevlartest opened the pull request:
Prototype code (frontend only):

- Volunteer Registration
![Screenshot 2020-08-18 at 10 43 06 AM](https://user-images.githubusercontent.com/68110443/90498225-ef2b4100-e13f-11ea-8be0-a9f1a6e5317d.png)
- Isolate Registration
![Screenshot 2020-08-18 at 10 43 37 AM](https://user-images.githubusercontent.com/68110443/90498220-edfa1400-e13f-11ea-9d8f-564272ac9634.png)

- Volunteer Matching
![Screenshot 2020-08-18 at 10 43 23 AM](https://user-images.githubusercontent.com/68110443/90498223-ee92aa80-e13f-11ea-8929-8bdf40a79646.png)
![Screenshot 2020-08-18 at 10 43 27 AM](https://user-images.githubusercontent.com/68110443/90498221-ee92aa80-e13f-11ea-92fe-db621c14bd9c.png)

- Isolate Request
![Screenshot 2020-08-18 at 10 44 07 AM](https://user-images.githubusercontent.com/68110443/90498215-ed617d80-e13f-11ea-87bd-482d6f28b415.png)

kevlartest committed:
Add skeleton Google Cloud Maven project

kevlartest committed:
Add skeleton registration pages structure

kevlartest committed:
Add Apache copyright headers

kevlartest committed:
Add HTML registration form

thomasogara committed:
Include registration form using jsp

thomasogara committed:
Add registration form

thomasogara committed:
Update registration link on the homepage

kevlartest committed:
Add isolate matching page

kevlartest committed:
Add volunteer match pages

thomasogara committed:
add information to isolate registration page

thomasogara committed:
adding request posting link to home page

thomasogara committed:
add ticket template

thomasogara committed:
add request form html artefact

thomasogara committed:
add request posting page

thomasogara committed:
Add TicketFactory

thomasogara committed:
add request and ticket logic

thomasogara committed:
update links on home page

thomasogara committed:
move request posting page

thomasogara committed:
move request form

thomasogara committed:
move ticket template

thomasogara committed:
separate DOM-related code into an ES module

thomasogara committed:
update TicketFactory to cache ticket template

thomasogara committed:
separate ticket-related code into an ES module

thomasogara committed:
add main.js to manage imports from ES modules

thomasogara committed:
update registration form to match volunteer form

thomasogara committed:
removing registration.html

thomasogara committed:
change placeholder to 'isolation address'

thomasogara committed:
change placeholder to 'isolation address'

thomasogara committed:
change wording on registration page

thomasogara committed:
change wording on registration page

thomasogara committed:
Merge pull request #4 from googleinterns/isolate_registration

Isolate Registration Feature

thomasogara committed:
Merge pull request #4 from googleinterns/isolate_registration

Isolate Registration Feature

thomasogara committed:
Merge branch 'devbase' into isolate_request_posting

thomasogara committed:
Merge branch 'devbase' into isolate_request_posting

thomasogara committed:
Merge pull request #3 from googleinterns/isolate_request_posting

Isolate Request Posting

thomasogara committed:
Merge pull request #3 from googleinterns/isolate_request_posting

Isolate Request Posting

thomasogara committed:
Merge branch 'devbase' into matching

thomasogara committed:
Merge branch 'devbase' into matching

thomasogara committed:
Merge pull request #2 from googleinterns/matching

Volunteer-Isolate Matching

thomasogara committed:
Merge pull request #2 from googleinterns/matching

Volunteer-Isolate Matching

thomasogara committed:
Merge pull request #1 from googleinterns/volunteerregistration

Volunteer Registration

thomasogara committed:
Merge pull request #1 from googleinterns/volunteerregistration

Volunteer Registration

thomasogara committed:
changing request ticket structure after feedback

thomasogara committed:
changing request ticket structure after feedback

magurmach commented on src/main/webapp/isolate/matcher.html:
```diff
@@ -0,0 +1,28 @@
+</head>
+<body>
+<p>You have been assigned a volunteer!</p>
+<p><b>Mark</b> will be helping you between <b>12:00 PM</b> and <b>2:00 PM</b>!</p>
```
Is the name being static intended here?

gjrb replied:
Indeed, even if this is the prototype, might be a good idea to add a TODO so you don't forget to remove it.

gjrb replied:
https://google.github.io/styleguide/htmlcssguide.html#Action_Items

thomasogara replied:
Sorry for the delay in replying. For the prototype, some elements were hardcoded, so this is intended

thomasogara replied:
fixed this now https://github.com/googleinterns/step226-2020/pull/5/commits/fa8e33b0a15eae6fbbb5b8f4e95c7ae7831a8403

magurmach commented on src/main/webapp/isolate/request/scripts/TicketFactory.js:
```diff
@@ -0,0 +1,112 @@
+      if (!response.ok) {
+        this.cacheTicketTemplate();
+      }
+      /* If the request succeeds, attempt to parse the body of the response */
```
Doesn't seems like best way to comment. Is the else unnecessary? Can't you just return from if?

thomasogara replied:
Is there any chance you could give some pointers as to what about the comment could be improved?

What do you mean exactly by 'return from if'? I'm not sure I can see myself how to refactor the conditional, so as to eliminate the 'else' on line 68?

gjrb replied:
I think what is meant is why do you need the else here at all? If you just deleted the else and moved the rest back, would the functionality be any different, is the else necessary at all?

gjrb replied:
I guess due to this function being recursive the else is necessary, so you don't run this code when the one which finally works returns.

gjrb replied:
Perhaps, this would be much easier code to read if it wasn't recursive?

thomasogara replied:
I've refactored the caching process to be iterative rather than recursive 👍 would it be possible to get some feedback, concerning whether the readability has been improved?

thomasogara replied:
relevant commits:

https://github.com/googleinterns/step226-2020/pull/5/commits/1afb5e1671fbe1e555d6e60e1f87cc88a85cbe10 https://github.com/googleinterns/step226-2020/pull/5/commits/132b135ab8d202acb0f00f70b0c54d1f4ab61a8b

magurmach commented on src/main/webapp/isolate/request/scripts/TicketFactory.js:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    const ticket = document.createElement('fieldset');
+    ticket.innerHTML = this.ticketTemplate;
+
+    /* Rename each ticket's input element */
+    for (let child of ticket.children) {
+      /*
+       * The name of the tickets always resembles a 2D array. This ensures that
+       * the browser can format the request properly, such that all of the
+       * tickets are recieved on the backend within a single POST request.
+       */
+      child.name = `tickets[${this.ticketCount}][${child.name}]`;
+    }
+
+    this.ticketCount++;
+    return ticket;
```
What is the purpose of this code?

thomasogara replied:
It creates a ticket for insertion into the request form. The renaming of the tickets is necessary to ensure that, on posting, the elements of each ticket (title, description etc.) are grouped on a per-ticket basis.

magurmach commented on src/main/webapp/isolate/request/scripts/tickets.js:
```diff
@@ -0,0 +1,79 @@
+    (ticket) => {
+      const TYPE_MISMATCH_PREFIX = `Ticket to be removed from form must not be`;
+
+      if (ticket === null) {
+        throw (TypeError(`${TYPE_MISMATCH_PREFIX} null.`));
+      }
+
+      else if (typeof ticket === 'undefined') {
+        throw (TypeError(`${TYPE_MISMATCH_PREFIX} undefined.`));
+      }
```
why there is this space between if and else if?

thomasogara replied:
I've fixed this now https://github.com/googleinterns/step226-2020/pull/5/commits/4f0530b45e59094d94ed6c539fa231e58cd2ab97

thomasogara committed:
remove 'else' from 'else if'

Semantically, there is no change in the behaviour of the code, as both if statements will cause immediate exit. However, seperating if / else if blocks across lines isn't allowed as per the style guide

gjrb commented on src/main/webapp/isolate/home.html:
```diff
@@ -0,0 +1,28 @@
+<body>
+  <p><a href="registration.jsp">Registration</a></p>
+  <p><a href="matcher.html">Matcher</a></p>
+  <p><a href="./request/index.jsp">Submit a request</a></p>
```
Nit: I don't think you need the ./ for subdirectories.

thomasogara committed:
remove unnecessary ./ from relative url

thomasogara committed:
remove unnecessary ./ from relative url

thomasogara committed:
remove unnecessary ./ from relative url

thomasogara committed:
remove unnecessary ./ from relative url

gjrb commented on src/main/webapp/isolate/request/scripts/DOMUtils.js:
```diff
@@ -0,0 +1,70 @@
+  const INVALID_PARENT_ELEMENT_MESSAGE =
+      `${ERROR_PREFIX} has an invalid 'parentElement' attribute`;
+
+  // Ensure that the element is not null.
```
Nit: obvious comment as you can tell exactly what is happening looking at the If statement below.

There are a couple of these in this file.

thomasogara replied:
fixed 👍 https://github.com/googleinterns/step226-2020/pull/5/commits/1740217e01e7df28db0f862de8f34dc31bbdbe6a

gjrb commented on src/main/webapp/isolate/request/scripts/TicketFactory.js:
```diff
@@ -0,0 +1,112 @@
+
+      /* If the request fails, try again. */
+      if (!response.ok) {
+        this.cacheTicketTemplate();
```
How do you break out of this? It seems like this is potentially a never ending recursive loop? What happens if this always fails?

thomasogara replied:
Oh, line 49 has a break condition, where 10 failed fetch attempts will cause an error to be thrown. This error is then caught, and an alert is presented to the user warning them that the template can't be fetched.

gjrb replied:
Oh I totally missed that!

gjrb commented on src/main/webapp/isolate/request/scripts/main.js:
```diff
@@ -0,0 +1,48 @@
+ * necessary members.
+ */
+
+import {addTicket, deleteTicket} from './tickets.js';
```
Do you need the ./ ?

gjrb commented on src/main/webapp/isolate/request/scripts/tickets.js:
```diff
@@ -0,0 +1,79 @@
+    const ticket = await ticketFactory.createTicket();
+    return ticket;
+  } catch (error) {
+    /* Pass the error up to the caller. */
```
Nit: obvious comment

thomasogara committed:
reformat caching logic to improve readability

thomasogara committed:
remove redundant comment

thomasogara committed:
remove redundant comments

thomasogara committed:
refactor caching to be iterative

thomasogara committed:
refactor caching

thomasogara committed:
add TODO to hard-coded prototype page

thomasogara committed:
resolve merge conflict in isolate/home.html

thomasogara committed:
Revert "remove unnecessary ./ from relative url"

This reverts commit dbd8e6a59f41cc761446f2580fa82c055b17033e.

thomasogara committed:
Revert "remove unnecessary ./ from relative url"

This reverts commit 2aeee8d4f3c1f46225c27ace8baea9c4eb90e25d.

gjrb commented on src/main/webapp/isolate/request/scripts/TicketFactory.js:
```diff
@@ -0,0 +1,110 @@
+      return false;
+    }
+    /* The ticketTemplate has already been cached. Report success. */
+    return true;
```
I think this would be more readable if you put this step at the top, and instead of checking that the template is null, check that it isn't null. That way I can clearly see that for a cached ticket the result is returned early.

You can then restructure the rest of the code to say, if the !response.ok return false, otherwise just execute the code for getting and caching the ticket.

thomasogara replied:
I've restructured the code now so that else statements are implicit, and so that the short-circuiting behaviour when the ticketTemplate is already cached is more easily understood 👍 Thanks for the guidance in terms of how to improve this piece of code https://github.com/googleinterns/step226-2020/pull/5/commits/441cbbdf3e92f89730e9c21e871e1dd3fb9c4eed https://github.com/googleinterns/step226-2020/pull/5/commits/7ba487129fa20c0a83c9299655ca6a27c29293f0

thomasogara committed:
refactor caching

thomasogara committed:
fix logic in caching function
